File: grindx/judges/python/lru-cache.py
"""Judge for LRU Cache — design class pattern."""

import importlib.util
import os
import random
import sys
from collections import OrderedDict

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from _common import load_cases, report_ac, report_wa


class ReferenceLRU:
    """Minimal correct LRU for generating expected values in large cases."""

    def __init__(self, capacity: int) -> None:
        self.cap = capacity
        self.cache: OrderedDict[int, int] = OrderedDict()

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        self.cache.move_to_end(key)
        return self.cache[key]

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache.move_to_end(key)
            self.cache[key] = value
        else:
            if len(self.cache) >= self.cap:
                self.cache.popitem(last=False)
            self.cache[key] = value


def generate_large_cases() -> list[dict]:
    """Generate 200K-operation case for TLE detection. Deterministic via seed."""
    random.seed(42)
    capacity = 1000
    num_ops = 200_000

    ref = ReferenceLRU(capacity)
    operations: list[str] = []
    op_inputs: list[list[int]] = []
    expected: list[int | None] = []

    for _ in range(num_ops):
        # Bias toward put: 0 = get, 1 or 2 = put
        if random.randint(0, 2) == 0:
            key = random.randint(1, 5000)
            result = ref.get(key)
            operations.append("get")
            op_inputs.append([key])
            expected.append(result)
        else:
            key = random.randint(1, 5000)
            value = random.randint(1, 100_000)
            ref.put(key, value)
            operations.append("put")
            op_inputs.append([key, value])
            expected.append(None)

    return [
        {
            "input": [capacity],
            "operations": operations,
            "op_inputs": op_inputs,
            "expected": expected,
            "category": "tle",
        }
    ]


def run(solution_path: str) -> None:
    spec = importlib.util.spec_from_file_location("solution", solution_path)
    if spec is None or spec.loader is None:
        print(f"Could not load solution: {solution_path}", file=sys.stderr)
        sys.exit(2)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)

    tc = load_cases("lru-cache")
    basic_cases = tc["cases"]
    large_cases = generate_large_cases()
    all_cases = basic_cases + large_cases
    total = len(all_cases)

    for i, case in enumerate(all_cases):
        capacity = case["input"][0]
        operations = case["operations"]
        op_inputs = case["op_inputs"]
        expected_vals = case["expected"]

        cache = mod.LRUCache(capacity)

        for j, op in enumerate(operations):
            if op == "put":
                cache.put(op_inputs[j][0], op_inputs[j][1])
            elif op == "get":
                result = cache.get(op_inputs[j][0])
                if expected_vals[j] is not None and result != expected_vals[j]:
                    # Build a descriptive case dict for the error report
                    err_case = {
                        "input": [f"op {j}: get({op_inputs[j][0]})"],
                        "expected": expected_vals[j],
                    }
                    report_wa(i, err_case, result, total)

    report_ac(total)


if __name__ == "__main__":
    run(sys.argv[1])

File: grindx/judges/javascript/lru-cache.js
/**
 * Judge for LRU Cache — design class pattern.
 */

const path = require("path");
const { loadCases, reportAc, reportWa, loadSolution, truncate } = require("./_common");

/**
 * Reference LRU implementation for generating expected output on large cases.
 */
class ReferenceLRU {
  constructor(capacity) {
    this.capacity = capacity;
    this.map = new Map();
  }
  get(key) {
    if (!this.map.has(key)) return -1;
    const val = this.map.get(key);
    this.map.delete(key);
    this.map.set(key, val);
    return val;
  }
  put(key, value) {
    if (this.map.has(key)) this.map.delete(key);
    this.map.set(key, value);
    if (this.map.size > this.capacity) {
      this.map.delete(this.map.keys().next().value);
    }
  }
}

function generateLargeCases() {
  const cases = [];
  const capacity = 1000;
  const numOps = 100000;
  let seed = 42;
  const rand = () => {
    seed = (seed * 1103515245 + 12345) & 0x7fffffff;
    return seed;
  };

  const operations = [];
  const opInputs = [];
  const expected = [];
  const ref = new ReferenceLRU(capacity);

  for (let i = 0; i < numOps; i++) {
    if (rand() % 3 === 0) {
      // ~33% get operations
      const key = rand() % 2000;
      operations.push("get");
      opInputs.push([key]);
      expected.push(ref.get(key));
    } else {
      // ~67% put operations
      const key = rand() % 2000;
      const val = rand() % 100000;
      operations.push("put");
      opInputs.push([key, val]);
      expected.push(null);
      ref.put(key, val);
    }
  }

  cases.push({
    input: [capacity],
    operations,
    op_inputs: opInputs,
    expected,
    category: "tle",
  });
  return cases;
}

function run(solutionPath) {
  // Class declarations are block-scoped in VM contexts and don't become
  // sandbox properties, so we wrap the solution in an IIFE that returns
  // the class explicitly.
  const fs = require("fs");
  const vm = require("vm");
  const code = fs.readFileSync(path.resolve(solutionPath), "utf-8");
  const wrapped = `(function() {\n${code}\nreturn { LRUCache };\n})()`;
  const result = vm.runInNewContext(wrapped, { require });
  const LRUCache = result.LRUCache;
  if (typeof LRUCache !== "function") {
    process.stderr.write("Solution does not export a LRUCache class\n");
    process.exit(2);
  }

  const tc = loadCases("lru-cache");
  const basicCases = tc.cases;
  const largeCases = generateLargeCases();
  const allCases = basicCases.concat(largeCases);
  const total = allCases.length;

  for (let i = 0; i < allCases.length; i++) {
    const c = allCases[i];
    const capacity = c.input[0];
    const ops = c.operations;
    const opInputs = c.op_inputs;
    const expectedResults = c.expected;

    const cache = new LRUCache(capacity);

    for (let j = 0; j < ops.length; j++) {
      const op = ops[j];
      const args = opInputs[j];
      let result;

      if (op === "get") {
        result = cache.get(args[0]);
      } else if (op === "put") {
        cache.put(args[0], args[1]);
        result = null;
      }

      if (expectedResults[j] !== null && result !== expectedResults[j]) {
        // Build a minimal preview for the failing operation
        const failInfo = {
          input: c.input,
          expected: `op[${j}] ${op}(${args.join(", ")}) => ${expectedResults[j]}`,
        };
        const wa = {
          verdict: "WA",
          failed_case: i,
          input_preview: truncate(JSON.stringify(failInfo.input)),
          expected_preview: truncate(failInfo.expected),
          actual_preview: truncate(JSON.stringify(result)),
          passed: i,
          total,
          category: c.category || "",
        };
        console.log(JSON.stringify(wa));
        process.exit(1);
      }
    }
  }
  reportAc(total);
}

run(process.argv[2]);
